pikelet/src/surface/lex.rs

```diff
@@ -0,0 +1,190 @@
+}
+
+/// The subset of `LexToken`s for the parser.
+/// The tokens in `LexToken`s which are excluded from this enum are:
+/// * Whitespace -- skipped.
+/// * EOF -- turned into None, rather than a token.
+/// * Error -- turned into Some(Err(...)).
+#[derive(Debug, Clone)]
+pub enum Token<'a> {
+    Colon,
+    Comma,
+    Fun,
+    DArrow,
+    Arrow,
+    LParen,
+    RParen,
+    LBrack,
+    RBrack,
+    LBrace,
+    RBrace,
+    RecordTerm,
+    RecordType,
+    Equal,
+    Dot,
+    CharLit(&'a str),
+    StrLit(&'a str),
+    Number(&'a str),
+    Name(&'a str),
+    Shift(&'a str),
+}
+
+impl<'a> Display for Token<'a> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        use Token as T;
+        match self {
+            T::Colon => write!(f, ":"),
+            T::Comma => write!(f, ","),
+            T::Fun => write!(f, "Fun"),
+            T::DArrow => write!(f, "=>"),
+            T::Arrow => write!(f, "->"),
+            T::LParen => write!(f, "("),
+            T::RParen => write!(f, ")"),
+            T::LBrack => write!(f, "["),
+            T::RBrack => write!(f, "]"),
+            T::LBrace => write!(f, "{{"),
+            T::RBrace => write!(f, "}}"),
+            T::RecordTerm => write!(f, "record"),
+            T::RecordType => write!(f, "Record"),
+            T::Equal => write!(f, "="),
+            T::Dot => write!(f, "."),
+            T::CharLit(s) => write!(f, "CharLit({})", s),
+            T::StrLit(s) => write!(f, "StrLit({})", s),
+            T::Number(s) => write!(f, "Number({})", s),
+            T::Name(s) => write!(f, "Write({})", s),
+            T::Shift(s) => write!(f, "Shift({})", s),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct LexicalError(std::ops::Range<usize>, &'static str);
+
+impl Display for LexicalError {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "Lexical error: {:?} {}", self.0, self.1)
+    }
+}
+
+pub struct LexIterator<'a>(logos::Lexer<LexToken, &'a str>);
+pub type Spanned<Tok, Loc, Error> = Result<(Loc, Tok, Loc), Error>;
+
+impl<'a> LexIterator<'a> {
+    pub fn new(source: &'a str) -> LexIterator<'a> {
+        LexIterator(LexToken::lexer(source))
+    }
+}
+
+impl<'a> Iterator for LexIterator<'a> {
+    type Item = Spanned<Token<'a>, usize, LexicalError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        use LexToken as LT;
+        use Token as T;
+        let lex = &mut self.0;
+
+        while let LT::Whitespace = &lex.token {
+            lex.advance();
+        }
+
+        const fn tok<'a>(
+            r: std::ops::Range<usize>,
+            t: Token<'a>,
+        ) -> Option<Spanned<Token<'a>, usize, LexicalError>> {
+            Some(Ok((r.start, t, r.end)))
+        }
+
+        let r = lex.range();
+        // I'm really not fond of this, but trying to split this match statement up
+        // appears to be more trouble than it is worth.
+        #[rustfmt::skip]
+        let it: Option<Self::Item> = match &lex.token {
```

Comment:
Perhaps:
```rust
let range = lexer.range();
#[rustfmt::skip]
let token = loop {
    match &lex.token {
        LexerToken::EOF => break None,
        LexerToken::Error => break Some(Err(LexicalError(range, "Lexical error"))),
        LexerToken::Whitespace => continue,
        LexerToken::Colon  => break tok(range, Token::Colon),
        ⋮
    }
};
lexer.advance();
token

Reply:
Had to make that { lexer.advance(); continue}, but other than that,
I removed the rustfmt::skip, because it didn't format terribly.